File: graphview/src/main/java/de/blox/graphview/energy/FruchtermanReingoldAlgorithm.java
package de.blox.graphview.energy;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Random;

import de.blox.graphview.Algorithm;
import de.blox.graphview.Edge;
import de.blox.graphview.EdgeRenderer;
import de.blox.graphview.Graph;
import de.blox.graphview.Node;
import de.blox.graphview.Vector;

/**
 *
 */
public class FruchtermanReingoldAlgorithm implements Algorithm {
    public static final int DEFAULT_ITERATIONS = 1000;
    public static final int CLUSTER_PADDING = 100;
    private static final double EPSILON = 0.0001D;
    private static final long SEED = 401678L;
    private final int iterations;
    private EdgeRenderer edgeRenderer = new EnergyEdgeRenderer();
    private Map<Node, Vector> disps = new HashMap<>();
    private Random rand = new Random(SEED);
    private int width;
    private int height;
    private float k;
    private float t;
    private float attraction_k;
    private float repulsion_k;

    public FruchtermanReingoldAlgorithm() {
        this(DEFAULT_ITERATIONS);
    }

    public FruchtermanReingoldAlgorithm(int iterations) {
        this.iterations = iterations;
    }

    private static int randInt(Random rand, int min, int max) {
        // nextInt is normally exclusive of the top value,
        // so add 1 to make it inclusive
        return rand.nextInt((max - min) + 1) + min;
    }

    private void randomize(List<Node> nodes) {

        for (Node node : nodes) {
            // create meta data for each node
            disps.put(node, new Vector());
            node.setPos(new Vector(randInt(rand, 0, width / 2), randInt(rand, 0, height / 2)));
        }
    }

    private void cool(int currentIteration) {
        t *= (1.0 - currentIteration / (double) iterations);
    }

    private void limitMaximumDisplacement(List<Node> nodes) {
        for (Node v : nodes) {
            float dispLength = (float) Math.max(EPSILON, getDisp(v).length());
            v.setPos(v.getPosition().add(getDisp(v).divide(dispLength).multiply((float) Math.min(dispLength, t))));
        }
    }

    private void calculateAttraction(List<Edge> edges) {
        for (Edge e : edges) {
            final Node v = e.getSource();
            final Node u = e.getDestination();

            Vector delta = v.getPosition().subtract(u.getPosition());
            float deltaLength = (float) Math.max(EPSILON, delta.length());
            setDisp(v, getDisp(v).subtract(delta.divide(deltaLength).multiply(forceAttraction(deltaLength))));
            setDisp(u, getDisp(u).add((delta.divide(deltaLength).multiply(forceAttraction(deltaLength)))));
        }
    }

    private void calculateRepulsion(List<Node> nodes) {
        for (Node v : nodes) {
            for (Node u : nodes) {
                if (!u.equals(v)) {
                    Vector delta = v.getPosition().subtract(u.getPosition());
                    float deltaLength = (float) Math.max(EPSILON, delta.length());
                    setDisp(v, getDisp(v).add(delta.divide(deltaLength).multiply(forceRepulsion(deltaLength))));
                }
            }
        }
    }

    private float forceAttraction(float x) {
        return (x * x) / attraction_k;
    }

    private float forceRepulsion(float x) {
        return (repulsion_k * repulsion_k) / x;
    }

    private Vector getDisp(Node node) {
        return disps.get(node);
    }

    private void setDisp(Node node, Vector disp) {
        disps.put(node, disp);
    }

    @Override
    public void run(Graph graph) {
        final int size = findBiggestSize(graph) * graph.getNodeCount();
        width = size;
        height = size;

        final List<Node> nodes = graph.getNodes();
        final List<Edge> edges = graph.getEdges();

        t = (float) (0.1 * Math.sqrt(width / 2 * height / 2));
        k = (float) (0.75 * Math.sqrt(width * height / nodes.size()));

        attraction_k = 0.75f * k;
        repulsion_k = 0.75f * k;

        randomize(nodes);

        for (int i = 0; i < iterations; i++) {
            calculateRepulsion(nodes);

            calculateAttraction(edges);

            limitMaximumDisplacement(nodes);

            cool(i);

            if (done()) {
                break;
            }
        }

        positionNodes(graph);
    }

    private void positionNodes(Graph graph) {
        Vector offset = getOffset(graph);
        List<Node> nodesVisited = new ArrayList<>();
        List<NodeCluster> nodeClusters = new ArrayList<>();
        for (Node node : graph.getNodes()) {
            node.setPos(new Vector(node.getX() - offset.getX(), node.getY() - offset.getY()));
        }

        for (Node node : graph.getNodes()) {
            if (nodesVisited.contains(node)) {
                continue;
            }

            nodesVisited.add(node);
            NodeCluster cluster = findClusterOf(nodeClusters, node);
            if (cluster == null) {
                cluster = new NodeCluster();
                cluster.add(node);
                nodeClusters.add(cluster);
            }

            followEdges(graph, cluster, node, nodesVisited);
        }

        positionCluster(nodeClusters);
    }

    private void positionCluster(List<NodeCluster> nodeClusters) {
        combineSingleNodeCluster(nodeClusters);

        NodeCluster cluster = nodeClusters.get(0);
        // move first cluster to 0,0
        cluster.offset(-cluster.rect.left, -cluster.rect.top);

        for (int i = 1; i < nodeClusters.size(); i++) {
            final NodeCluster nextCluster = nodeClusters.get(i);
            final float xDiff = nextCluster.rect.left - cluster.rect.right - CLUSTER_PADDING;
            final float yDiff = nextCluster.rect.top - cluster.rect.top;
            nextCluster.offset(-xDiff, -yDiff);
            cluster = nextCluster;
        }
    }


    private void combineSingleNodeCluster(List<NodeCluster> nodeClusters) {
        NodeCluster firstSingleNodeCluster = null;
        final Iterator<NodeCluster> iterator = nodeClusters.iterator();
        while (iterator.hasNext()) {
            NodeCluster cluster = iterator.next();
            if (cluster.size() == 1) {
                if (firstSingleNodeCluster == null) {
                    firstSingleNodeCluster = cluster;
                    continue;
                }

                firstSingleNodeCluster.concat(cluster);
                iterator.remove();
            }
        }
    }

    private void followEdges(Graph graph, NodeCluster cluster, Node node, List<Node> nodesVisited) {
        for (Node successor : graph.successorsOf(node)) {
            if (nodesVisited.contains(successor)) {
                continue;
            }

            nodesVisited.add(successor);
            cluster.add(successor);

            followEdges(graph, cluster, successor, nodesVisited);
        }

        for (Node predecessor : graph.predecessorsOf(node)) {
            if (nodesVisited.contains(predecessor)) {
                continue;
            }

            nodesVisited.add(predecessor);
            cluster.add(predecessor);

            followEdges(graph, cluster, predecessor, nodesVisited);
        }
    }

    private NodeCluster findClusterOf(List<NodeCluster> clusters, Node node) {
        for (NodeCluster cluster : clusters) {
            if (cluster.contains(node)) {
                return cluster;
            }
        }

        return null;
    }

    private int findBiggestSize(Graph graph) {
        int size = 0;
        for (Node node : graph.getNodes()) {
            size = Math.max(size, Math.max(node.getHeight(), node.getWidth()));
        }

        return size;
    }

    private Vector getOffset(Graph graph) {
        float offsetX = Float.MAX_VALUE;
        float offsetY = Float.MAX_VALUE;
        for (Node node : graph.getNodes()) {
            offsetX = Math.min(offsetX, node.getX());
            offsetY = Math.min(offsetY, node.getY());
        }
        return new Vector(offsetX, offsetY);
    }

    private boolean done() {
        return t < 1.0 / Math.max(height, width);
    }

    @Override
    public void drawEdges(Canvas canvas, Graph graph, Paint linePaint) {
        edgeRenderer.render(canvas, graph, linePaint);
    }

    @Override
    public void setEdgeRenderer(EdgeRenderer renderer) {
        this.edgeRenderer = renderer;
    }

    private static class NodeCluster {
        private List<Node> nodes = new ArrayList<>();
        private RectF rect;

        public void add(Node node) {
            nodes.add(node);

            if (rect == null) {
                rect = new RectF(node.getX(), node.getY(), node.getX() + node.getWidth(), node.getY() + node.getHeight());
            } else {
                rect.left = Math.min(rect.left, node.getX());
                rect.top = Math.min(rect.top, node.getY());
                rect.right = Math.max(rect.right, node.getX() + node.getWidth());
                rect.bottom = Math.max(rect.bottom, node.getY() + node.getHeight());
            }
        }

        public boolean contains(Node node) {
            return nodes.contains(node);
        }

        public int size() {
            return nodes.size();
        }

        public void concat(NodeCluster cluster) {
            for (Node node : cluster.nodes) {
                node.setPos(new Vector(rect.right + CLUSTER_PADDING, rect.top));
                add(node);
            }
        }

        public void offset(float xDiff, float yDiff) {
            for (Node node : nodes) {
                node.setPos(node.getPosition().add(xDiff, yDiff));
            }

            rect.offset(xDiff, yDiff);
        }
    }
}

File: graphview/src/main/java/de/blox/graphview/energy/FruchtermanReingoldAlgorithm.kt
package de.blox.graphview.energy

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import de.blox.graphview.Edge
import de.blox.graphview.Graph
import de.blox.graphview.Layout
import de.blox.graphview.Node
import de.blox.graphview.edgerenderer.ArrowEdgeRenderer
import de.blox.graphview.edgerenderer.EdgeRenderer
import de.blox.graphview.util.Size
import de.blox.graphview.util.VectorF
import java.util.*
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

class FruchtermanReingoldAlgorithm @JvmOverloads constructor(private val iterations: Int = DEFAULT_ITERATIONS) :
        Layout {
    private var edgeRenderer: EdgeRenderer = ArrowEdgeRenderer()
    private val disps: MutableMap<Node, VectorF> = HashMap()
    private val rand = Random(SEED)
    private var width: Int = 0
    private var height: Int = 0
    private var k: Float = 0.toFloat()
    private var t: Float = 0.toFloat()
    private var attraction_k: Float = 0.toFloat()
    private var repulsion_k: Float = 0.toFloat()

    private fun randomize(nodes: List<Node>) {
        nodes.forEach { node ->
            // create meta data for each node
            disps[node] = VectorF()
            node.setPosition(
                    randInt(rand, 0, width / 2).toFloat(),
                    randInt(rand, 0, height / 2).toFloat()
            )
        }
    }

    private fun cool(currentIteration: Int) {
        t *= (1.0f - currentIteration / iterations.toFloat())
    }

    private fun limitMaximumDisplacement(nodes: List<Node>) {
        nodes.forEach {
            val dispLength = max(EPSILON, getDisp(it).length().toDouble()).toFloat()
            it.setPosition(
                    it.position.add(
                            getDisp(it).divide(dispLength).multiply(
                                    min(
                                            dispLength,
                                            t
                                    )
                            )
                    )
            )
        }
    }

    private fun calculateAttraction(edges: List<Edge>) {
        edges.forEach { (v, u) ->
            val delta = v.position.subtract(u.position)
            val deltaLength = max(EPSILON, delta.length().toDouble()).toFloat()
            setDisp(
                    v,
                    getDisp(v).subtract(delta.divide(deltaLength).multiply(forceAttraction(deltaLength)))
            )
            setDisp(
                    u,
                    getDisp(u).add(delta.divide(deltaLength).multiply(forceAttraction(deltaLength)))
            )
        }
    }

    private fun calculateRepulsion(nodes: List<Node>) {
        nodes.forEach { v ->
            nodes.forEach { u ->
                if (u != v) {
                    val delta = v.position.subtract(u.position)
                    val deltaLength = max(EPSILON, delta.length().toDouble()).toFloat()
                    setDisp(
                            v,
                            getDisp(v).add(delta.divide(deltaLength).multiply(forceRepulsion(deltaLength)))
                    )
                }
            }
        }
    }

    private fun forceAttraction(x: Float): Float {
        return x * x / attraction_k
    }

    private fun forceRepulsion(x: Float): Float {
        return repulsion_k * repulsion_k / x
    }

    private fun getDisp(node: Node): VectorF {
        return disps.getValue(node)
    }

    private fun setDisp(node: Node, disp: VectorF) {
        disps[node] = disp
    }

    override fun run(graph: Graph, shiftX: Float, shiftY: Float): Size {
        val size = findBiggestSize(graph) * graph.nodeCount
        width = size
        height = size

        val nodes = graph.nodes
        val edges = graph.edges

        t = (0.1 * sqrt((width / 2f * height / 2f).toDouble())).toFloat()
        k = (0.75 * sqrt((width * height / nodes.size.toFloat()).toDouble())).toFloat()

        attraction_k = 0.75f * k
        repulsion_k = 0.75f * k

        randomize(nodes)

        (0 until iterations).forEach { i ->
            calculateRepulsion(nodes)

            calculateAttraction(edges)

            limitMaximumDisplacement(nodes)

            cool(i)

            if (done()) {
                return@forEach
            }
        }

        positionNodes(graph)

        shiftCoordinates(graph, shiftX, shiftY)

        return calculateGraphSize(graph)
    }

    private fun shiftCoordinates(graph: Graph, shiftX: Float, shiftY: Float) {
        graph.nodes.forEach { node ->
            node.setPosition(VectorF(node.x + shiftX, node.y + shiftY))
        }
    }

    private fun positionNodes(graph: Graph) {
        val (x, y) = getOffset(graph)
        val nodesVisited = ArrayList<Node>()
        val nodeClusters = ArrayList<NodeCluster>()
        graph.nodes.forEach { node ->
            node.setPosition(VectorF(node.x - x, node.y - y))
        }

        graph.nodes.forEach { node ->
            if (nodesVisited.contains(node)) {
                return@forEach
            }

            nodesVisited.add(node)
            var cluster = findClusterOf(nodeClusters, node)
            if (cluster == null) {
                cluster = NodeCluster()
                cluster.add(node)
                nodeClusters.add(cluster)
            }

            followEdges(graph, cluster, node, nodesVisited)
        }

        positionCluster(nodeClusters)
    }

    private fun positionCluster(nodeClusters: MutableList<NodeCluster>) {
        combineSingleNodeCluster(nodeClusters)

        var cluster = nodeClusters[0]
        // move first cluster to 0,0
        cluster.offset(-cluster.rect.left, -cluster.rect.top)

        for (i in 1 until nodeClusters.size) {
            val nextCluster = nodeClusters[i]
            val xDiff = nextCluster.rect.left - cluster.rect.right - CLUSTER_PADDING.toFloat()
            val yDiff = nextCluster.rect.top - cluster.rect.top
            nextCluster.offset(-xDiff, -yDiff)
            cluster = nextCluster
        }
    }

    private fun combineSingleNodeCluster(nodeClusters: MutableList<NodeCluster>) {
        var firstSingleNodeCluster: NodeCluster? = null
        val iterator = nodeClusters.iterator()
        while (iterator.hasNext()) {
            val cluster = iterator.next()
            if (cluster.size() == 1) {
                if (firstSingleNodeCluster == null) {
                    firstSingleNodeCluster = cluster
                    continue
                }

                firstSingleNodeCluster.concat(cluster)
                iterator.remove()
            }
        }
    }

    private fun followEdges(
            graph: Graph,
            cluster: NodeCluster,
            node: Node,
            nodesVisited: MutableList<Node>
    ) {
        graph.successorsOf(node).forEach { successor ->
            if (nodesVisited.contains(successor)) {
                return@forEach
            }

            nodesVisited.add(successor)
            cluster.add(successor)

            followEdges(graph, cluster, successor, nodesVisited)
        }

        graph.predecessorsOf(node).forEach { predecessor ->
            if (nodesVisited.contains(predecessor)) {
                return@forEach
            }

            nodesVisited.add(predecessor)
            cluster.add(predecessor)

            followEdges(graph, cluster, predecessor, nodesVisited)
        }
    }

    private fun findClusterOf(clusters: List<NodeCluster>, node: Node): NodeCluster? {
        return clusters.firstOrNull { it.contains(node) }
    }

    private fun findBiggestSize(graph: Graph): Int {
        return graph.nodes
                .map { max(it.height, it.width) }
                .max()
                ?: 0
    }

    private fun getOffset(graph: Graph): VectorF {
        var offsetX = java.lang.Float.MAX_VALUE
        var offsetY = java.lang.Float.MAX_VALUE
        graph.nodes.forEach { node ->
            offsetX = min(offsetX, node.x)
            offsetY = min(offsetY, node.y)
        }
        return VectorF(offsetX, offsetY)
    }

    private fun done(): Boolean {
        return t < 1.0 / max(height, width)
    }

    override fun drawEdges(canvas: Canvas, graph: Graph, linePaint: Paint) {
        edgeRenderer.render(canvas, graph, linePaint)
    }

    override fun setEdgeRenderer(renderer: EdgeRenderer) {
        this.edgeRenderer = renderer
    }

    private fun calculateGraphSize(graph: Graph): Size {

        var left = Integer.MAX_VALUE
        var top = Integer.MAX_VALUE
        var right = Integer.MIN_VALUE
        var bottom = Integer.MIN_VALUE
        for (node in graph.nodes) {
            left = min(left.toFloat(), node.x).toInt()
            top = min(top.toFloat(), node.y).toInt()
            right = max(right.toFloat(), node.x + node.width).toInt()
            bottom = max(bottom.toFloat(), node.y + node.height).toInt()
        }

        return Size(right - left, bottom - top)
    }

    private class NodeCluster {
        val nodes = arrayListOf<Node>()
        var rect: RectF = RectF()

        fun add(node: Node) {
            nodes.add(node)

            if (nodes.size == 1) {
                rect.apply {
                    left = node.x
                    top = node.y
                    right = node.x + node.width
                    bottom = node.y + node.height
                }
            } else {
                rect.apply {
                    left = min(left, node.x)
                    top = min(top, node.y)
                    right = max(right, node.x + node.width)
                    bottom = max(bottom, node.y + node.height)
                }
            }
        }

        operator fun contains(node: Node): Boolean {
            return nodes.contains(node)
        }

        fun size(): Int {
            return nodes.size
        }

        fun concat(cluster: NodeCluster) {
            cluster.nodes.forEach { node ->
                node.setPosition(VectorF(rect.right + CLUSTER_PADDING, rect.top))
                add(node)
            }
        }

        fun offset(xDiff: Float, yDiff: Float) {
            nodes.forEach { node ->
                node.setPosition(node.position.add(xDiff, yDiff))
            }

            rect.offset(xDiff, yDiff)
        }
    }

    companion object {
        const val DEFAULT_ITERATIONS = 1000
        const val CLUSTER_PADDING = 100
        private const val EPSILON = 0.0001
        private const val SEED = 401678L

        private fun randInt(rand: Random, min: Int, max: Int): Int {
            // nextInt is normally exclusive of the top value,
            // so add 1 to make it inclusive
            return rand.nextInt(max - min + 1) + min
        }
    }
}
